idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/folding/kinds.kt

```diff
@@ -1,58 +1,70 @@
 package arrow.meta.ide.plugins.proofs.folding
 
-import arrow.meta.ide.dsl.utils.typeProjections
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.dsl.utils.getType
 import arrow.meta.ide.dsl.utils.typeReferences
+import arrow.meta.phases.ExtensionPhase
 import com.intellij.lang.ASTNode
-import com.intellij.lang.folding.FoldingBuilderEx
 import com.intellij.lang.folding.FoldingDescriptor
 import com.intellij.openapi.editor.Document
-import com.intellij.psi.PsiElement
-import com.intellij.psi.impl.source.SourceTreeToPsiMap
+import com.intellij.psi.util.strictParents
 import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtTypeProjection
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.KtUserType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.types.KotlinType
 
-class KindFoldingBuilder : FoldingBuilderEx() {
-  override fun getPlaceholderText(node: ASTNode): String? =
-    SourceTreeToPsiMap.treeElementToPsi(node)?.let { psiElement ->
-      (psiElement as KtElement).typeProjections
-        .filter { !it.text.startsWith("Kind") }
-        .map { it.text }
+val IdeMetaPlugin.codeFoldingOnKinds: ExtensionPhase
+  get() = registerFoldingBuilder(
+    foldingBuilder(
+      placeHolderText = { node: ASTNode ->
+        (node.psi as KtElement).typeReferences
```

Comment:
is it possible to do something like (node.psi as KtTypeReference)?
Same in other places.